src/dune_rules/melange_rules.ml

```diff
@@ -29,21 +56,38 @@ let js_includes ~loc ~sctx ~target_dir ~requires_link ~scope ~js_ext =
           let lib_name = Lib.name lib in
           let lib = local_of_lib ~loc lib in
           let info = Lib.Local.info lib in
-          let lib_dir = Lib_info.src_dir info in
-          let dst_dir = lib_output_dir ~target_dir ~lib_dir in
-          let open Memo.O in
-          let modules_group =
-            Dir_contents.get sctx ~dir:lib_dir
-            >>= Dir_contents.ocaml
-            >>| Ml_sources.modules ~for_:(Library lib_name)
-          in
-          let* source_modules = modules_group >>| Modules.impl_only in
-          let of_module m =
-            let output = make_js_name ~js_ext ~dst_dir m in
-            Dep.file (Path.build output)
-          in
-          Resolve.Memo.return
-            (List.map source_modules ~f:of_module |> Dep.Set.of_list)
+          match Lib_info.virtual_ info with
+          | Some _ ->
+            (* Js rules for virtual libs are added by their implementations *)
```

Comment:
This was surprising, but it turns out that when calculating `Lib.Compile.requires_link` for an implementation library in `add_rules_for_libraries`

https://github.com/ocaml/dune/blob/ef472e94e3ab76a4f2e63d1d7043f77cec5319ab/src/dune_rules/melange_rules.ml#L194-L195

the virtual library is returned as part of the list in the resulting value 🤔

Reply:
That sounds wrong. Virtual libraries are supposed to be absent from the linking closure.

Reply:
I took a look, but I am not sure how to create a test that proofs this. Do you have any ideas, or existing tests that I could use as reference?

Reply:
No way to write a test but you can add an assert inside the function that calculates the closure.